biggraphite/drivers/elasticsearch.py

```diff
@@ -207,6 +209,80 @@ class InvalidArgumentError(Error, bg_accessor.InvalidArgumentError):
     """Callee did not follow requirements on the arguments."""
 
 
+def _parse_wildcard_component(component):
+    """Given a complex component, this builds a wildcard constraint."""
+    value = ""
+    for subcomponent in component:
+        if isinstance(subcomponent, bg_glob.AnySequence):
+            value += "*"
+        elif isinstance(subcomponent, six.string_types):
+            value += subcomponent
+        elif isinstance(subcomponent, bg_glob.AnyChar):
+            value += '?'
+        else:
+            raise Error("Unhandled type '%s'" % subcomponent)
+    return value
+
+
+def _parse_regexp_component(component):
+    """Given a complex component, this builds a regexp constraint."""
+    if isinstance(component, bg_glob.Globstar):
+        return ".*"
+
+    regex = ""
+    for subcomponent in component:
+        if isinstance(subcomponent, bg_glob.AnySequence):
+            regex += "[^.]*"
+        elif isinstance(subcomponent, six.string_types):
+            regex += subcomponent
+        elif isinstance(subcomponent, bg_glob.CharNotIn):
+            regex += '[^' + ''.join(subcomponent.values) + ']'
+        elif isinstance(subcomponent, bg_glob.CharIn):
+            regex += '[' + ''.join(subcomponent.values) + ']'
+        elif isinstance(subcomponent, bg_glob.SequenceIn):
+            if subcomponent.negated:
+                regex += '[^.]*'
+            else:
+                regex += '(' + '|'.join(subcomponent.values) + ')'
+        elif isinstance(subcomponent, bg_glob.AnyChar):
+            regex += '[^.]'
+        else:
+            raise Error("Unhandled type '%s'" % subcomponent)
+    return regex
+
+
+def parse_complex_component(component):
+    """Given a complex component, this builds a constraint."""
+    if all([
+            any([
+                 isinstance(sub_c, bg_glob.AnySequence),
+                 isinstance(sub_c, bg_glob.AnyChar),
+                 isinstance(sub_c, six.string_types),
+            ]) for sub_c in component
+            ]):
+        return 'wildcard', _parse_wildcard_component(component)
+    return 'regexp', _parse_regexp_component(component)
+
+
+def parse_simple_component(component):
+    """Given a component with a simple type, this builds a constraint."""
+    value = component[0]
+    if isinstance(value, bg_glob.AnySequence):
+        return None, None  # No constrain
+    elif isinstance(value, six.string_types):
+        return 'term', value
+    elif isinstance(value, bg_glob.CharNotIn):
+        return 'regexp', '[^' + ''.join(value.values) + ']'
+    elif isinstance(value, bg_glob.CharIn):
+        return 'regexp', '[' + ''.join(value.values) + ']'
+    elif isinstance(value, bg_glob.SequenceIn):
+        return 'terms', value.values
+    elif isinstance(value, bg_glob.AnyChar):
+        return 'wildcard', '?'
+    else:
+        raise Error("Unhandled type '%s'" % value)
+
```

Comment:
can you start adding unit tests for these (it's easy as they don't depend on es).

if you want to implement them in a follow up review to avoid conflicts, add a TODO

Reply:
Done